File: exercism/javascript/circular-buffer/circular-buffer.js

class CircularBuffer {
  constructor (size) {
    this.clear(size)
  }
  clear (size) {
    size = size || this.buffer.length
    this.buffer = Array(size)
    this.writePointer = 0
    this.readPointer = 0
    this.semaphore = 0
  }
  read () {
    if (!this.semaphore) throw new BufferEmptyException
    const val = this.buffer[this.readPointer]
    this.readPointer = this.next(this.readPointer)
    this.semaphore -= 1
    return val
  }
  write (val) {
    if (!val) return
    if (this.semaphore >= this.buffer.length) throw new BufferFullException()
    this.buffer[this.writePointer] = val
    this.writePointer = this.next(this.writePointer)
    this.semaphore += 1
  }
  forceWrite (val) {
    try {
      this.write(val)
    } catch (BufferFullException) {
      this.buffer[this.writePointer] = val
      this.writePointer = this.next(this.writePointer)
      this.readPointer = this.next(this.readPointer)
    }
  }
  next (pointer) {
    return (pointer + 1) % this.buffer.length
  }
}


class BufferEmptyException extends Error {}
class BufferFullException extends Error {}


module.exports = {
  circularBuffer: (n) => new CircularBuffer(n),
  bufferEmptyException: () => new BufferEmptyException(),
  bufferFullException: () => new BufferFullException()
}

File: exercism/python/circular-buffer/circular_buffer.py

class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, size):
        self.size = size
        self.clear()

    def clear(self):
        self.buffer = [None] * self.size
        self.wp = 0  # write pointer
        self.rp = 0  # read pointer
        self.readable = 0  # semaphore

    def read(self):
        if self.readable == 0:
            raise BufferEmptyException()
        val = self.buffer[self.rp]
        self.buffer[self.rp] = None
        self.rp = (self.rp + 1) % self.size
        self.readable -= 1
        return val

    def write(self, val):
        if self.readable == self.size:
            raise BufferFullException()
        self.buffer[self.wp] = val
        self.wp = (self.wp + 1) % self.size
        self.readable += 1

    def overwrite(self, val):
        if self.readable < self.size:
            return self.write(val)
        self.buffer[self.rp] = val
        self.rp = (self.rp + 1) % self.size
